Add support for relationships

sharplet opened the pull request.

sharplet committed:
Remove unnecessary TestResources module

sharplet committed:
Extract JSONAPI.Data parameter object

The `JSONAPI.Data` struct packages up the data needed for clients to
query the JSON document during decoding. With this in place, it will be
straight-forward to add support for relationships by adding a new
property to this type.

sharplet committed:
Add support for decoding has-one relationships

sharplet committed:
Rename HasOneChildResource to ChildResource

sharplet committed:
Add support for has-many relationships

sharplet committed:
Add test for deeper levels of linkage

sharplet committed:
Add error message when no matching relationship found

sharplet committed:
Stop pretending to mutate relationships while decoding

The code as written makes it appear that we're removing an object from
the "included" array whenever a match is found. This doesn't appear to
actually be the case: as we recurse the tree of relationships, each node
appears to be acting on its own copy of the relationships, making the
mutation unnecessary.

sharplet committed:
Don't expose public decode methods for JSONAPI.Data

sharplet committed:
Don't wrap rewrap relationships in fake document

Adds a new method `JSONAPI.Data.decodeResource(_:referencing:)`, which
allows the parsed "included" key to be passed around directly while
recursively decoding a tree of relationships.

sharplet committed:
Remove inaccurate use of "simple" in naming

gfontenot commented on Sources/ArgoJSONAPI/ArgoJSONAPI.swift:
```diff
@@ -1,20 +1,20 @@
   let json = JSON(object)
-
-  return (json <| "data") >>- Document<T>.decodeResource
+  return Document<T>.decodeResource(json)
```
We should be able to remove the `Document<T>` here, right?

sharplet replied:
Nah, the return type is `Decoded<T>`, not `Document<T>`

gfontenot replied:
Ah, I see. Misread, my bad.

gfontenot commented on Sources/ArgoJSONAPI/ArgoJSONAPI.swift:
```diff
@@ -1,20 +1,20 @@
 import Argo
-import Runes
+
+public enum JSONAPI {
```
What's the reasoning behind defining this here and then extending it elsewhere?

sharplet replied:
It's just declaring a namespace basically, which has to happen _somewhere_, and so I chose to put it in the file named after the library. Do you think it's worth adding a doc comment to indicate that it's only intended as a namespace?

gfontenot replied:
Ah, gotcha. I think a public comment would probably be good, yeah.

gfontenot reviewed (approved):
This is so awesome

sharplet committed:
Redeclare Argo operators

Operators normally aren't exported with an `@_exported import` or via
the umbrella header. This change redeclares Argo's operators using the
same precedence groups, removing the need to explicitly `import Argo`
when you `import ArgoJSONAPI`.

sharplet commented:
I think this is about done, but planning to wait for https://github.com/Quick/Quick/pull/711 to land before merging this.
